pebblo/app/service/local_ui/retriever_apps.py

```diff
@@ -0,0 +1,562 @@
+
+        sorted_document = self.sort_db_retrievals(
+            retrieval_data, "retrieved_from", all_apps
+        )
+        response = {}
+        for user_name, user_data in sorted_document.items():
+            accessed_time = []
+            for data in user_data:
+                accessed_time.append(parser.parse(data.get("prompt_time")))
+            response[user_name] = {
+                "retrievals": user_data,
+                "last_accessed_time": self.fetch_last_accessed_time(accessed_time),
+            }
+        return response
+
+    def prepare_db_retrieval_response(
+        self, retriever_app, prompt_details, total_prompt_with_findings
+    ):
+        all_retrievals = []
+        retriever_data = retriever_app.data
+        app_name = retriever_data.get("name")
+        prompt_details[app_name] = {}
+
+        # fetch total retrievals
+        _, ai_retrievals = self.db.query(
+            table_obj=AiRetrievalTable, filter_query={"app_name": app_name}
+        )
+        for ai_retrieval in ai_retrievals:
+            try:
+                all_retrievals.append(ai_retrieval.data)
+                retrieval_data = {"name": app_name}
+                retrieval_data.update(ai_retrieval.data)
+                self.total_retrievals.append(retrieval_data)
+                prompt_details, total_prompt_with_findings = self.get_prompt_details(
+                    prompt_details,
+                    ai_retrieval.data,
+                    app_name,
+                    total_prompt_with_findings,
+                )
+            except Exception as ex:
+                logger.warning(
+                    f"[Dashboard]:  Error while iterating retrieval for app {app_name} Error: {ex}"
+                )
+
+        # fetch active users per app
+        active_users = self.get_db_active_users(ai_retrievals, all_apps=True)
+        self.add_accumulate_active_users(active_users)
+
+        # fetch vector dbs per app
+        vector_dbs = self.get_vector_dbs(retriever_data.get("chains"))
+        self.retrieval_vectordbs.extend(vector_dbs)
+
+        # fetch documents name per app
+        documents = self.get_all_db_documents(ai_retrievals, all_apps=True)
+
+        app_details = RetrievalAppListDetails(
+            name=app_name,
+            owner=retriever_data.get("owner"),
+            retrievals=all_retrievals,
+            active_users=list(active_users.keys()),
+            vector_dbs=vector_dbs,
+            documents=list(documents.keys()),
+        )
+        return app_details.dict(), prompt_details, total_prompt_with_findings
+
+    @staticmethod
+    def _sort_retrievals_with_retrieval_count(retrievals: list) -> list:
+        """
+        Sort the list based on the retrieval count in the descending order
+        :param retrievals: retrievals list
+        :return:  sorted retrievals list
+        """
+        sorted_data = sorted(
+            retrievals, key=lambda item: len(item["retrievals"]), reverse=True
```

Comment:
Should we use get?

Reply:
yes, we are using `len()` and the apply `sorted()` on top of it.

Reply:
Updated.